### DISCUSSION TITLE
Excluding targets in task dependencies

### DISCUSSION THREAD

**Discussion by jbreckmckye** [vercel/turborepo#8597] [Help] [answered]
### Summary

Here's a simplified version of my context:

- I have a large monorepo that includes one `app`, several `features` and many `libraries`
- The `app` has static dependencies (libraries) and dynamic dependencies (features).
- Feature modules are logically isolated from the app that loads them. This means that (simplifying somewhat) we can trust that feature module changes don't require app unit tests to re-run.

What I want is a configuration where app unit tests **only run** if the app or its library dependencies change.

Here's a generalised example of the application structure

```mermaid
flowchart TD
  L1[Library 1]
  L2[Library 2]
  FM1[Feature module 1]
  App[App]

  L1 --> App
  L2 --> App
  L2 --> FM1
  FM1 -.-> App
```

So breaking down the behaviour, what we want is

| If this changes... | Run unit tests for...            |
|--------------------|----------------------------------|
| Library 1          | Library 1, App                   |
| Library 2          | Library 2, Feature module 1, App |
| Feature module 1   | Feature module 1                 |
| App                | App                              |

**Is there a clean way to achieve what I'm after?**

**Comment by anthonyshew** [comment, marked as answer]
Very interesting. I might not be thinking cleverly enough but I don't know of a way to do this with `turbo` today that doesn't create more problems than it solves.

The only thing I can think of is to do something like (pseudocoding):
- If the difference between SHA 1 and SHA 2 is _only_ in Feature Module 1
  - `turbo run unit-tests --filter=feature-module-1 --only`
- Else
  - `turbo run unit-tests`

You can see you'd be working outside of `turbo`-land here to get some of the logic you need.

I *think* the missing piece here would be getting to negate on an [arbitrary task relationship](https://turbo.build/repo/docs/reference/configuration#arbitrary-task-relationships). However, I don't know how palatable that is as a feature (I can see people shooting themselves in the foot with this), and we would need to see more hunger from the community to be able to prioritize it (I can't recall someone asking for this before).

**Reply by jbreckmckye** [reply]
Hi Anthony, thanks for the fast response!

Getting to grips with Turbo, I think I'm coming to the same conclusion. One thing I'd misunderstood was that although I can control the _execution_ graph (e.g. by creating different tasks for different kinds of modules), I can't selectively exclude modules from the 'change graph'.

From what I can tell (and I might be wrong here) the actual hash for invalidating / triggering a task includes not only the module's files, but also hashes based on its dependencies.

Therefore it isn't possible for me to update `Feature module 1` (in my example) without invaliding the unit tests for `App`, because just the act of changing a file in `dependency` alters the hash of the `dependent`. (I assume this is intentional behaviour by Turborepo)
